src/main/java/echoserver/EchoServer.java

```diff
@@ -19,15 +19,19 @@ public EchoServer(Listenable listener, Executor executor, Echoable echoer, Logga
     public void start() throws Exception {
         Connection socketConnection;
         while((socketConnection = this.listener.listen()) != null) {
-            Connection finalSocketConnection = socketConnection;
-            Runnable runnable = () -> {
-                try {
-                    this.echoer.echo(finalSocketConnection);
-                } catch (IOException e) {
-                    this.logger.log("Failed to echo connection.");
-                }
-            };
+            Runnable runnable = runnable(socketConnection);
             executor.execute(runnable);
         }
     }
+
+    private Runnable runnable(Connection connection) {
+        Runnable runnable = () -> {
+            try (Connection socketConnection = connection) {
+                this.echoer.echo(connection);
+            } catch (IOException e) {
+                this.logger.log("Failed to read the socket connection.");
```

Comment:
What if the exception is raised when we write to the socket? I think we need a better way to provide the error context to the person reading the logs.

Reply:
What about something like this: 
this.logger.log("Failed to read from or write to the socket connection. " + e)